### DISCUSSION TITLE
hCaptcha validation with event hooks/middleware

### DISCUSSION THREAD

**Discussion by XaFigg** [pocketbase/pocketbase#3663] [Q&A] [answered]
Hello,
I would like to implement hCaptcha on the frontend side to avoid abuse.
On the other hand, you have to validate the captcha on the backend side, and to do this I want to use event hooks/middleware in Go.
I would like to restrict this validation only on certain collections.

For event hooks, I don't know if there is a way to cancel the request, for example in the `OnRecordBeforeCreateRequest` event.
For the middleware, I haven't been able to find a way to specify it on an existing route, or am I not doing it the right way?

Thanks to those who will help me ^^

**Comment by ganigeorgiev** [comment]
You can cancel requests from a hook by simply returning a Go error (_if you want a formatted error response you can return [apis.ApiError](https://pocketbase.io/docs/go-routing/#error-response)_).

For the middleware, you can register a global middleware and check the request path of the current request context to target only specific route(s) (eg. with `c.Path()` or with the resolved path params `c.Request().URL.Path`; note that for the latter the collection route can be accessed both by id and name).

Both approaches are fine but the `*Request` hook one would be probably easier.

_Side-note: builtin captcha support will be eventually implemented in the future as part of https://github.com/pocketbase/pocketbase/issues/731 and https://github.com/pocketbase/pocketbase/issues/1232._

**Reply by XaFigg** [reply]
Thank you for your answer.
Indeed I also think that the approach with event hooks is simpler.
What is strange is that from the pocketbase admin panel, I receive an error when creating the error (with in my case a 401 forbidden), but the record is still created. Maybe because of the admin status?

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not sure that I understand what do you mean. I forgot to mention that the Admin UI uses the same web APIs so you'll have to exclude the requests made by admins. This can be done by inspect the [request auth context](https://pocketbase.io/docs/go-routing/#retrieving-the-current-auth-state):

```go
app.OnRecordBeforeCreateRequest("yourCollection").Add(func(e *core.RecordCreateEvent) error {
	admin, _ := e.HttpContext.Get(apis.ContextAdminKey).(*models.Admin)
	if admin != nil {
		return nil // ignore for admins
	}

	// perform your captcha validations...
	if invalidCaptcha {
		return apis.NewForbiddenError("Invalid captcha", nil)
	}

	return nil
})
```

**Reply by XaFigg** [reply]
In fact I didn't read it correctly lol... I wasn't returning the correct function. But with the code you provided, everything works fine, thank you!
Have a good evening/day
